### ISSUE TITLE
Parse error for .exenv-version file

### ISSUE THREAD

**Issue by efueger** [fazibear/codeclimate-credo#10]
## Description

Customer reported today: 
- Credo engine started giving him the error below
- excluding the `.exenv-version` file in his .codeclimate.yml allows builds to pass.

----

```
info: Some source files could not be parsed correctly and are excluded:

   1) .exenv-version

{"type":"Issue","remediation_points":100000,"location":{"path":"apps/questions/lib/questions/auth/strategies/somos/authenticate.ex","lines":{"end":85,"begin":85}},"description":"Pipe chain should start with a raw value.","content":{"body":"Pipes (`|>`) can can become more readable by starting with a \"raw\" value.\n\nSo while this is easily comprehendable:\n\n    list\n    |> Enum.take(5)\n    |> Enum.shuffle\n    |> pick_winner()\n\nThis might be harder to read:\n\n    Enum.take(list, 5)\n    |> Enum.shuffle\n    |> pick_winner()\n\nAs always: This is just a suggestion. Check the configuration options for\ntweaking or disabling this check.\n"},"check_name":"pipe_chain_start","categories":["Style"]}
```

**Comment by caironoleto** [comment]
Actually, the error is not in `.exenv-version` file but is something related to engine's output.

I fixed the error in line 85, as the error output (it's my project. Thanks, @efueger) and build passed.

**Comment by fazibear** [comment]
So the problem is resolved ?

**Comment by caironoleto** [comment]
Nops. If there's any lint error, the build fails.

**Comment by fazibear** [comment]
@caironoleto Can you provide me some example. I can't reproduce it your problem.

**Comment by caironoleto** [comment]
Another example: I changed our codeclimate file with the stricted option, and then the output is:

![image](https://user-images.githubusercontent.com/9125/31317285-789ca938-ac14-11e7-8a07-29e24e92dce0.png)

```
Issue unparseable: info: Some source files could not be parsed correctly and are excluded:

   1) .exenv-version

{"type":"Issue","remediation_points":100000,"location":{"path":"apps/questions/test/questions/contexts/mocks_users_test.exs","lines":{"end":52,"begin":52}},"description":"Large numbers should be written with underscores: 1_064_000","content":{"body":"Numbers can contain underscores for readability purposes.\nThese do not affect the value of the number, but can help read large numbers\nmore easily.\n\n    141592654 # how large is this number?\n\n    141_592_654 # ah, it's in the hundreds of millions!\n\nLike all `Readability` issues, this one is not a technical concern.\nBut you can improve the odds of others reading and liking your code by making\nit easier to follow.\n"},"check_name":"large_numbers","categories":["Style"]}
``` 

Our codeclimate file:

```yml
---
engines:
  eslint:
    enabled: true
    channel: stable
    extensions:
    - .js
    - .jsx
  fixme:
    enabled: true
  credo:
    enabled: true
    channel: "beta"
    checks:
      module_doc:
        enabled: false
    config:
      strict: true
ratings:
  paths:
  - "**.css"
  - "**.js"
  - "**.jsx"
  - "**.ex"
  - "**.exs"
exclude_paths:
- config/
- apps/react/node_modules/
- apps/bo/node_modules/
- _build/
- deps/
```

**Comment by fazibear** [comment]
I've tested it with `codeclimate-cli` and it's working fine. There is a warning, but the output is parsed fine because all warnings go to stderr. 

Is `codeclimate-cli` works for your project ?

**Comment by fazibear** [comment]
Anyway I've updated this engine. @efueger Can you deploy new version ?

**Comment by efueger** [comment]
@fazibear  Sure - will let you know once it's deployed.

**Comment by efueger** [comment]
Hi @caironoleto and @fazibear - I think there are still a couple of things to work through with this one.

@caironoleto
For context, when providing a branch comparison, we actually run two analyses - one of the branch head, and one of the branch merge base (the commit at which the branch started). If either, or even one of these analyses fail, we're unable to provide a comparison.

- In your case, I do see a few builds that contained the `.exenv-version` exclusion, and their builds passed. 
- But, the merge base still errored because its .codeclimate.yml didn't contain that exclusion.

1. Can you try merging that change to your default branch?
2. Once you have that "good" commit, you can [rebase](https://docs.codeclimate.com/docs/fixing-branches-and-pull-requests#section-instructions-for-rebasing) any affected open branches.

----

@fazibear 
I cloned the Credo repo, ran `codeclimate engines:install` (to make sure I had the newest image for Credo), but it's giving the same complaint (about a different file).

- Excluding the file does allow the build to pass.
- Here's [my repo](https://github.com/efueger/credo).
- Can you test that out on your CLI?

Here's the error message I see:
```
Running credo: Done!
error: (JSON::ParserError) 743: unexpected token at 'info: Some source files could not be parsed correctly and are excluded:

   1) .template.check.ex

{"type":"Issue","remediation_points":100000,"location":{"path":"lib/credo/cli/output/issues_short_list.ex","lines":{"end":1,"begin":1}},"description":"Modules should have a @moduledoc tag.","content":{"body":"Every module should contain comprehensive documentation.\n\n    # preferred\n\n    defmodule MyApp.Web.Search do\n      @moduledoc \"\"\"\n      This module provides a public API for all search queries originating\n      in the web layer.\n      \"\"\"\n    end\n\n    # also okay: explicitly say there is no documentation\n\n    defmodule MyApp.Web.Search do\n      @moduledoc \"\"\"\n      This module provides a public API for all search queries originating\n      in the web layer.\n      \"\"\"\n    end\n\nMany times a sentence or two in plain english, explaining why the module\nexists, will suffice. Documenting your train of thought this way will help\nboth your co-workers and your future-self.\n\nOther times you will want to elaborate even further and show some\nexamples of how the module's functions can and should be used.\n\nIn some cases however, you might not want to document things about a module,\ne.g. it is part of a private API inside your project. Since Elixir prefers\nexplicitness over implicit behaviour, you should \"tag\" these modules with\n\n    @moduledoc false\n\nto make it clear that there is no intention in documenting it.\n"},"check_name":"module_doc","categories":["Clarity"]}'
```

**Comment by fazibear** [comment]
@efueger 

I've done following steps:

- checkout this repo: https://github.com/efueger/credo
- build a docker from master: https://github.com/fazibear/codeclimate-credo with `docker build -t codeclimate/codeclimate-credo .`
- comment `channel: beta` in `.codeclimate.yml`
- run `codeclimate analyze --dev `

Looks like everything is working! This is an output:

```
Starting analysis
Running eslint: Done!
Running fixme: Done!
Running credo: Done!

== .credo.exs (1 issue) ==
72: TODO found [fixme]

== README.md (5 issues) ==
151: TODO found [fixme]
306: TODO found [fixme]
306: TODO found [fixme]
422: TODO found [fixme]
422: FIXME found [fixme]

== lib/credo/check/consistency/exception_names/collector.ex (2 issues) ==
43: TODO found [fixme]
43: Found a TODO tag in a comment: # TODO: how is this `case` necessary [credo]

== lib/credo/check/consistency/space_around_operators.ex (8 issues) ==
27: TODO found [fixme]
27: Found a TODO tag in a comment: # TODO: add *ignored* operators, so you can add "|" and still write [credo]
29: Found a FIXME tag in a comment: # FIXME: this seems to be already implemented, but there don't seem to be [credo]
29: FIXME found [fixme]
74: Found a TODO tag in a comment: # TODO: Consider moving these checks inside the Collector. [credo]
74: TODO found [fixme]
103: Found a TODO tag in a comment: # TODO: this implementation is a bit naive. improve it. [credo]
103: TODO found [fixme]

== lib/credo/check/design/alias_usage.ex (3 issues) ==
92: Function takes too many parameters (arity is 7, max is 5). [credo]
96: Function takes too many parameters (arity is 7, max is 5). [credo]
115: Function takes too many parameters (arity is 7, max is 5). [credo]

== lib/credo/check/design/duplicated_code.ex (3 issues) ==
63: Function body is nested too deep (max depth is 2, was 4). [credo]
215: Found a TODO tag in a comment: # TODO: Put in AST helper [credo]
215: TODO found [fixme]

== lib/credo/check/design/tag_fixme.ex (5 issues) ==
2: Found a FIXME tag in a comment: FIXME comments are used to indicate places where source code needs fixing.

Example:

    # FIXME: this does no longer work, research new API url
    defp fun do
      # ...
    end

The premise here is that FIXME should indeed be fixed as soon as possible and
are therefore reported by Credo.

Like all `Software Design` issues, this is just advice and might not be
applicable to your project/situation.
 [credo]
3: FIXME found [fixme]
7: FIXME found [fixme]
12: FIXME found [fixme]
26: FIXME found [fixme]

== lib/credo/check/design/tag_todo.ex (5 issues) ==
2: Found a TODO tag in a comment: TODO comments are used to remind yourself of source code related things.

Example:

    # TODO: move this to a Helper module
    defp fun do
      # ...
    end

The premise here is that TODO should be dealt with in the near future and
are therefore reported by Credo.

Like all `Software Design` issues, this is just advice and might not be
applicable to your project/situation.
 [credo]
3: TODO found [fixme]
7: TODO found [fixme]
12: TODO found [fixme]
26: TODO found [fixme]

== lib/credo/check/find_lint_attributes.ex (2 issues) ==
59: TODO found [fixme]
59: Found a TODO tag in a comment: # TODO: warn that a new lint attribute was read while one was still active [credo]

== lib/credo/check/lint_attribute.ex (4 issues) ==
3: TODO found [fixme]
3: Found a TODO tag in a comment: # TODO: remove these [credo]
4: TODO found [fixme]
4: Found a TODO tag in a comment: # TODO: as they are for debug purposes only [credo]

== lib/credo/check/readability/large_numbers.ex (2 issues) ==
122: Pipe chain should start with a raw value. [credo]
128: Pipe chain should start with a raw value. [credo]

== lib/credo/check/readability/max_line_length.ex (2 issues) ==
35: TODO found [fixme]
35: Found a TODO tag in a comment: # TODO v1.0: this should be two different params [credo]

== lib/credo/check/refactor/append_single_item.ex (2 issues) ==
32: File has the variable name before the pattern while most of the files have the variable name after the pattern when naming parameter pattern matches [credo]
37: File has the variable name before the pattern while most of the files have the variable name after the pattern when naming parameter pattern matches [credo]

== lib/credo/check/refactor/cyclomatic_complexity.ex (2 issues) ==
59: Duplicate code found in lib/credo/check/refactor/perceived_complexity.ex:59 (mass: 58). [credo]
107: Duplicate code found in lib/credo/check/refactor/perceived_complexity.ex:107 (mass: 52). [credo]

== lib/credo/check/refactor/perceived_complexity.ex (2 issues) ==
59: Duplicate code found in lib/credo/check/refactor/cyclomatic_complexity.ex:59 (mass: 58). [credo]
107: Duplicate code found in lib/credo/check/refactor/cyclomatic_complexity.ex:107 (mass: 52). [credo]

== lib/credo/check/refactor/pipe_chain_start.ex (1 issue) ==
45: TODO found [fixme]

== lib/credo/check/warning/name_redeclaration_by_case.ex (1 issue) ==
101: Function is too complex (CC is 10, max is 9). [credo]

== lib/credo/check/warning/name_redeclaration_by_def.ex (3 issues) ==
40: TODO found [fixme]
40: Found a TODO tag in a comment: # TODO: make customizable via params [credo]
109: Function is too complex (CC is 11, max is 9). [credo]

== lib/credo/check/warning/name_redeclaration_by_fn.ex (1 issue) ==
97: Function is too complex (CC is 10, max is 9). [credo]

== lib/credo/check/warning/unused_file_operation.ex (1 issue) ==
5: TODO found [fixme]

== lib/credo/check/warning/unused_function_return_helper.ex (2 issues) ==
251: TODO found [fixme]
251: Found a TODO tag in a comment: # TODO: move to AST helper? [credo]

== lib/credo/check/warning/unused_keyword_operation.ex (2 issues) ==
5: TODO found [fixme]
19: Duplicate code found in lib/credo/check/warning/unused_tuple_operation.ex:19, lib/credo/check/warning/unused_list_operation.ex:19 (mass: 61). [credo]

== lib/credo/check/warning/unused_list_operation.ex (2 issues) ==
5: TODO found [fixme]
19: Duplicate code found in lib/credo/check/warning/unused_tuple_operation.ex:19, lib/credo/check/warning/unused_keyword_operation.ex:19 (mass: 61). [credo]

== lib/credo/check/warning/unused_path_operation.ex (2 issues) ==
5: TODO found [fixme]
20: Duplicate code found in lib/credo/check/warning/unused_regex_operation.ex:20 (mass: 63). [credo]

== lib/credo/check/warning/unused_regex_operation.ex (2 issues) ==
5: TODO found [fixme]
20: Duplicate code found in lib/credo/check/warning/unused_path_operation.ex:20 (mass: 63). [credo]

== lib/credo/check/warning/unused_tuple_operation.ex (2 issues) ==
5: TODO found [fixme]
19: Duplicate code found in lib/credo/check/warning/unused_keyword_operation.ex:19, lib/credo/check/warning/unused_list_operation.ex:19 (mass: 61). [credo]

== lib/credo/cli/command/explain.ex (2 issues) ==
14: TODO found [fixme]
14: Found a TODO tag in a comment: # TODO: explain used exec options [credo]

== lib/credo/cli/command/suggest.ex (2 issues) ==
84: TODO found [fixme]
84: Found a TODO tag in a comment: # TODO: offer short list (?) [credo]

== lib/credo/cli/output/categories.ex (2 issues) ==
37: TODO found [fixme]
37: FIXME found [fixme]

== lib/credo/cli/output/explain.ex (11 issues) ==
46: Function takes too many parameters (arity is 7, max is 5). [credo]
50: Function takes too many parameters (arity is 7, max is 5). [credo]
159: Function body is nested too deep (max depth is 2, was 3). [credo]
185: TODO found [fixme]
229: TODO found [fixme]
229: Found a TODO tag in a comment: # TODO: format things in backticks in help texts [credo]
267: Space missing after comma [credo]
292: Space missing after comma [credo]
300: Space missing after comma [credo]
313: There are spaces around operators most of the time, but not here. [credo]
325: There are spaces around operators most of the time, but not here. [credo]

== lib/credo/cli/output/issue_helper.ex (1 issue) ==
74: Function takes too many parameters (arity is 6, max is 5). [credo]

== lib/credo/cli/output/issues_by_scope.ex (1 issue) ==
117: Function body is nested too deep (max depth is 2, was 3). [credo]

== lib/credo/cli/output/summary.ex (1 issue) ==
70: There are spaces around operators most of the time, but not here. [credo]

== lib/credo/code/module.ex (2 issues) ==
230: TODO found [fixme]
230: Found a TODO tag in a comment: # TODO: write unit test [credo]

== lib/credo/code/sigils.ex (1 issue) ==
17: Pipe chain should start with a raw value. [credo]

== lib/credo/code/strings.ex (2 issues) ==
12: TODO found [fixme]
12: Found a TODO tag in a comment: # TODO v1.0: this should not remove heredocs, since [credo]

== lib/credo/execution/task.ex (2 issues) ==
42: TODO found [fixme]
42: Found a TODO tag in a comment: # TODO: improve message [credo]

== lib/credo/exs_loader.ex (2 issues) ==
35: TODO found [fixme]
35: Found a TODO tag in a comment: # TODO: support regex modifiers [credo]

== test/credo/check/design/tag_fixme_test.exs (10 issues) ==
30: TODO found [fixme]
30: FIXME found [fixme]
32: FIXME found [fixme]
32: FIXME found [fixme]
57: FIXME found [fixme]
59: FIXME found [fixme]
62: FIXME found [fixme]
63: FIXME found [fixme]
65: FIXME found [fixme]
67: FIXME found [fixme]

== test/credo/check/design/tag_helper_test.exs (18 issues) ==
16: TODO found [fixme]
26: TODO found [fixme]
31: TODO found [fixme]
34: TODO found [fixme]
35: TODO found [fixme]
46: TODO found [fixme]
51: TODO found [fixme]
54: TODO found [fixme]
55: TODO found [fixme]
72: TODO found [fixme]
85: TODO found [fixme]
87: TODO found [fixme]
90: TODO found [fixme]
91: TODO found [fixme]
93: TODO found [fixme]
95: TODO found [fixme]
97: TODO found [fixme]
101: TODO found [fixme]

== test/credo/check/design/tag_todo_test.exs (30 issues) ==
51: TODO found [fixme]
51: TODO found [fixme]
55: TODO found [fixme]
59: TODO found [fixme]
73: FIXME found [fixme]
73: TODO found [fixme]
75: TODO found [fixme]
75: TODO found [fixme]
89: FIXME found [fixme]
92: TODO found [fixme]
106: TODO found [fixme]
106: TODO found [fixme]
109: FIXME found [fixme]
123: TODO found [fixme]
123: TODO found [fixme]
126: FIXME found [fixme]
139: FIXME found [fixme]
139: TODO found [fixme]
141: TODO found [fixme]
154: TODO found [fixme]
163: TODO found [fixme]
172: TODO found [fixme]
182: TODO found [fixme]
204: TODO found [fixme]
206: TODO found [fixme]
206: TODO found [fixme]
209: TODO found [fixme]
210: TODO found [fixme]
212: TODO found [fixme]
214: TODO found [fixme]

== test/credo/check/readability/large_numbers_test.exs (3 issues) ==
125: Large numbers should be written with underscores: 50_000 [credo]
189: Pipe chain should start with a raw value. [credo]
200: Pipe chain should start with a raw value. [credo]

== test/credo/check/readability/string_sigils_test.exs (9 issues) ==
38: Pipe chain should start with a raw value. [credo]
44: Pipe chain should start with a raw value. [credo]
50: Pipe chain should start with a raw value. [credo]
56: Pipe chain should start with a raw value. [credo]
62: Pipe chain should start with a raw value. [credo]
68: Pipe chain should start with a raw value. [credo]
74: Pipe chain should start with a raw value. [credo]
84: Pipe chain should start with a raw value. [credo]
90: Pipe chain should start with a raw value. [credo]

== test/credo/code/charlists_test.exs (6 issues) ==
9: TODO found [fixme]
10: TODO found [fixme]
14: TODO found [fixme]
17: TODO found [fixme]
22: TODO found [fixme]
26: TODO found [fixme]

== test/credo/code/heredocs_test.exs (9 issues) ==
9: TODO found [fixme]
12: TODO found [fixme]
13: TODO found [fixme]
16: TODO found [fixme]
23: TODO found [fixme]
24: TODO found [fixme]
27: TODO found [fixme]
38: TODO found [fixme]
47: TODO found [fixme]

== test/credo/code/strings_test.exs (7 issues) ==
9: TODO found [fixme]
12: TODO found [fixme]
13: TODO found [fixme]
16: TODO found [fixme]
24: TODO found [fixme]
38: TODO found [fixme]
47: TODO found [fixme]

== test/credo/priority_test.exs (13 issues) ==
35: There are spaces around operators most of the time, but not here. [credo]
36: There are spaces around operators most of the time, but not here. [credo]
37: There are spaces around operators most of the time, but not here. [credo]
38: There are spaces around operators most of the time, but not here. [credo]
39: There are spaces around operators most of the time, but not here. [credo]
40: There are spaces around operators most of the time, but not here. [credo]
68: There are spaces around operators most of the time, but not here. [credo]
69: There are spaces around operators most of the time, but not here. [credo]
70: There are spaces around operators most of the time, but not here. [credo]
71: There are spaces around operators most of the time, but not here. [credo]
72: There are spaces around operators most of the time, but not here. [credo]
73: There are spaces around operators most of the time, but not here. [credo]
74: There are spaces around operators most of the time, but not here. [credo]

== test/credo/sources_test.exs (10 issues) ==
13: Pipe chain should start with a raw value. [credo]
26: Pipe chain should start with a raw value. [credo]
40: Pipe chain should start with a raw value. [credo]
49: Pipe chain should start with a raw value. [credo]
64: Pipe chain should start with a raw value. [credo]
80: Pipe chain should start with a raw value. [credo]
95: Pipe chain should start with a raw value. [credo]
105: Pipe chain should start with a raw value. [credo]
120: Pipe chain should start with a raw value. [credo]
168: Pipe chain should start with a raw value. [credo]

== test/support/test_application.exs (2 issues) ==
20: There are spaces around operators most of the time, but not here. [credo]
20: There are spaces around operators most of the time, but not here. [credo]

Analysis complete! Found 213 issues.
```

**Comment by efueger** [comment]
@caironoleto - a new version of the Credo engine was deployed yesterday.
- I've tested against my repo, and it seems to be happy now :)
- Let me know if you have any more trouble with that!

Thanks for your help throughout the troubleshooting process.

**Comment by caironoleto** [comment]
@efueger thank you. Everything is working right now! :)

**Comment by caironoleto** [comment]
@fazibear thank you too <3
